### ISSUE TITLE
Reordering of form fields?

### ISSUE THREAD

**Issue by hongkongkiwi** [kirankunigiri/zenstack-ui#2]
How can I reorder the form fields?

Right now, I'm reordering via the DB, but that seems a really not great way to do it once I hit production.

**Comment by kirankunigiri** [comment]
This is a good idea. The simplest way I can think of is having an extra order number attribute added to the field, which you set in the metadata object before passing it to the zenstack-ui config.

```ts
itemFields.name.order = 0;
itemFields.description.order = 1;
```

For the implementation, I can think of 2 options:

1. Automatically assign an order number to every field based on the current zmodel order to the metadata object. However, if you want to just quickly move some items to the top of the form, you would have to manually re-order (by setting the order attribute value) all of the fields, or use negative values, which can get confusing
2. By default, the order field is undefined. When a user sets an order number in the metadata object, all fields with order values are prioritized and placed at the top in the order of their order numbers. The ones with undefined order numbers are placed below in the order of the zmodel file.

I think option 2 makes more sense as it is easier to make quick changes without needing to re-order all the fields but let me know what you think, or if you have any other ideas as well. It should be a fairly easy code fix because all I have to do is run a sort on the fields.

**Comment by hongkongkiwi** [comment]
> Automatically assign an order number to every field based on the current zmodel order to the metadata object. However, if you want to just quickly move some items to the top of the form, you would have to manually re-order (by setting the order attribute value) all of the fields, or use negative values, which can get confusing

I see the only advantage of (1) is that it's pretty easy to understand for newcomers as all fields are numbered and have a clear order from ZS UI. The order is already pre-defined, and you can manipulate it how you want. As you say though, reordering a portion means dealing with the whole list which may be annoying as you need more code.

> By default, the order field is undefined. When a user sets an order number in the metadata object, all fields with order values are prioritized and placed at the top in the order of their order numbers. The ones with undefined order numbers are placed below in the order of the zmodel file.

All things considered; I think (2) is the best approach and I like it a lot. It's flexible and means you can just have a list of known fields and handle them. Then when an unknown (new) field appears it gets put at the end (or front) automatically by design.

My only suggestion is to have a flag which defines whether the unordered fields appear before or after the zordered fields. By default, the unordered fields show after ordered fields as you mention, but it's nice to be able to manipulate this (since if you have a long form, it would be nice to see any new or unhandled fields at the start).

For above idea I also considered negative zordering, but that gets dumb if all your fields are before the unordered as you need to order negatively backwards all the time. So just having a flag to indicate whether unordered appears before or after is way simpler.

**Comment by kirankunigiri** [comment]
This is now implemented in v0.0.21! Fields now have an order attribute. Also, a new options object has been implemented with a few settings, one of which is whether the ordered fields go first or last compared to the unordered fields.

```ts
// Ordering example
itemFields.name.order = 0;
itemFields.description.order = 1;

export const zsuiConfig: ZenStackUIConfigCreate = {
	// hooks, schemas, etc...
	options: {
		showOrderedFieldsFirst: false,
	},
};
```

You can also override these options for a specific form.
```ts
// Example to pass options override, uncomment to test
const optionsOverride: Partial<ZenStackUIOptions> = {
	showOrderedFieldsFirst: false,
	showUpdateButton: true,
};

return (
	<ZSUpdateForm {...props} optionsOverride={optionsOverride}/>
)
```
